src/main/java/undo/UndoController.java

```diff
@@ -0,0 +1,84 @@
+package undo;
+
+import backend.Logic;
+import backend.resource.TurboIssue;
+import javafx.application.Platform;
+import javafx.geometry.Insets;
+import javafx.scene.control.Label;
+import javafx.util.Pair;
+import ui.NotificationController;
+import ui.components.Notification;
+import undo.actions.Action;
+import util.DialogMessage;
+
+import java.util.Optional;
+
+public class UndoController {
+
+    @SuppressWarnings("SpellCheckingInspection")
+    private static final String OCTICON_INFO = "\uf059";
+
+    private final Logic logic;
+    private final NotificationController notificationController;
+    private Optional<Pair<TurboIssue, Action<TurboIssue>>> undoBuffer;
+
+    // Holds an undo buffer of size 1.
+    // Buffer stores the undo action.
+    public UndoController(Logic logic, NotificationController notificationController) {
+        this.logic = logic;
+        this.notificationController = notificationController;
+        undoBuffer = Optional.empty();
+    }
+
+    // Gets called when an action is undone
+    public void undoCallback() {
+        if (undoBuffer.isPresent()) {
+            determineLogicMethod(undoBuffer.get().getKey(), undoBuffer.get().getValue());
+            undoBuffer = Optional.empty();
+        }
+    }
+
+    // Adds an action to the undo buffer.
+    @SuppressWarnings("unchecked")
+    public void addAction(TurboIssue issue, Action action) {
+        determineLogicMethod(issue, action);
+        undoBuffer = Optional.of(new Pair<>((TurboIssue) action.act(issue), action.invert()));
+        Notification notification = new Notification(createInfoOcticon(),
+                "Undo " + action.getDescription() + " for #" + issue.getId() + ": " + issue.getTitle(),
+                "Undo", this::undoCallback);
+        notificationController.showNotification(notification);
+    }
+
+    // Check to determine which logic method to call
+    @SuppressWarnings("unchecked")
+    private void determineLogicMethod(TurboIssue issue, Action action) {
+        if (action.getType() == Action.ActionType.CHANGELABELS) {
+            logic.replaceIssueLabels(issue, action)
```

Comment:
Yeah, that wouldn't work, because it's overloading. The method to use will be decided at compile-time. There's no runtime decision based on object type.

I was thinking of this:

```java
private void determineLogicMethod(TurboIssue issue, Action<TurboIssue> action) {
    action.act(issue)
        .thenApply(success ->
            showErrorDialogOnFailure(success, issue, action));
}
```

which is the de facto way to decide what to do based on type. Was hoping to leverage that rather than reinvent it (and have to repeat it everywhere Actions are used, without the compiler helping us check exhaustiveness... which is my real concern).

We have to agree on what Actions represent: right now I'm thinking of them as the command object that we put on the undo stack, and you're thinking of them as the issue diff.

Since once we decide the type of command (i.e. labels, milestones, etc.), we know what type of diff to use, I'm more inclined to have Actions represent commands. The decision is really which resource we're dealing with. After `action.act(issue)` selects the type of command/Action, a specific type of diff can be applied in it without requiring further polymorphism. What do you think?